[PXCT-1437] Cloud - Provisioning documentation changes (#2645)
arduino/docs-content: content/arduino-cloud/03.cloud-interface/00.sketches/sketches.md

This file will be visible as a "Secret" tab in the Cloud Editor and is named `arduino_secrets.h`, which is not visible on the Cloud platform.

Note that if you are using the offline IDE / Arduino CLI, you will need to manually create this file. More information is in the **Offline Sketches section**.

> ## thingsProperties.h and the NetworkConfigurator

Reviewer: Not sure if this needs to be in 'code' for the linter to not register it as an issue

The thingProperties.h file plays a key role in managing credential capabilities, including the NetworkConfigurator, which enables two main features:

- **Credentials stored on NVS**: Boards can now securely store network settings in Non-Volatile Storage (NVS), removing them from the sketch (secrets.h).

> - **Over-the-Air (OTA) communication**: Enables the possibility to provide network configuration settings via Bluetooth (BLE).

Reviewer: ```suggestion
- **Over-the-Air (OTA) communication**: Enables the possibility to provide network configuration settings via Bluetooth LE.
```

> The `thingProperties.h` will be generated accordingly to the provisioning mechanism, so if the board has been registered using the provisioning 2.0, the `thingProperties.h` file will automatically have the NetworkConfigurator component enabled. A board registered with Provisioning 2.0 includes `Arduino_NetworkConfigurator.h` in the generated `thingProperties.h` file. For more information about device provisioning have a look [here.](https://docs.arduino.cc/arduino-cloud/hardware/device-provisioning/)

Reviewer: ```suggestion
The `thingProperties.h` will be generated accordingly to the provisioning mechanism, so if the board has been registered using the Provisioning 2.0, the `thingProperties.h` file will automatically have the NetworkConfigurator component enabled. A board registered with Provisioning 2.0 includes `Arduino_NetworkConfigurator.h` in the generated `thingProperties.h` file. For more information about device provisioning have a look [here.](https://docs.arduino.cc/arduino-cloud/hardware/device-provisioning/)
```

### How the NetworkConfigurator works

To work, the NetworkConfigurator needs:

- **One or more Configurator Agents**: Objects that handle the communication between the board and the user device (PC, laptop, or Mobile phone).
- **A Key-Value Storage library**: the NetworkConfigurator needs an external storage library that implements the KVStoreInterface. Arduino provides the `Arduino_KVStore` library for handling the storage and saving the NetworkConfigurator configurations.
- **A ConnectionHandler**: the object responsible for the board's Internet connection management.

The `NetworkConfigurator` library out-of-the-box provides two Agents:

- `BLEAgent` for handling the BLE communication.
- `SerialAgent` for the Serial communication.

## thingsProperties.h Default Setup

Here is how the `thingsProperties.h` file changes to set up the NetworkConfigurator. This setup is automatically generated if the board has been registered with Provisioning 2.0.

> *** Do not follow these steps if the board has not been registered with the Provisioning 2.0. For more information about device provisioning have a look [here.](https://docs.arduino.cc/arduino-cloud/hardware/device-provisioning/) ***

Reviewer: ```suggestion
*** Do not follow these steps if the board has not been registered with the Provisioning 2.0. For more information about device provisioning have a look [here](https://docs.arduino.cc/arduino-cloud/hardware/device-provisioning/) ***.
```

> ### Libraries and object declarations

Reviewer: ```suggestion
### Libraries and Object Declarations
```

```arduino
NetworkConfigurator.addAgent(BLEAgent);
NetworkConfigurator.addAgent(SerialAgent);
NetworkConfigurator.setStorage(kvStore);
ArduinoCloud.setConfigurator(NetworkConfigurator);
```

The final `thingProperties.h` file:

```arduino
#include <ArduinoIoTCloud.h>
#include <Arduino_ConnectionHandler.h>
#include <Arduino_NetworkConfigurator.h>
#include <configuratorAgents/agents/BLEAgent.h>
#include <configuratorAgents/agents/SerialAgent.h>
void onVariableChange();
int variable;
KVStore kvStore;
BLEAgentClass BLEAgent;
SerialAgentClass SerialAgent;
WiFiConnectionHandler ArduinoIoTPreferredConnection; 
NetworkConfiguratorClass NetworkConfigurator(ArduinoIoTPreferredConnection);
void initProperties(){
  NetworkConfigurator.addAgent(BLEAgent);
  NetworkConfigurator.addAgent(SerialAgent);
  NetworkConfigurator.setStorage(kvStore);
  ArduinoCloud.setConfigurator(NetworkConfigurator);
// For changing the default reset pin (2) uncomment and set your preferred pin. Use DISABLE_PIN for disabling the reset procedure.
// NetworkConfigurator.setReconfigurePin(YOUR_PIN);
  ArduinoCloud.addProperty(variable, READWRITE, ON_CHANGE, onVariableChange);
}
```

### Disable the BLEAgent

> To save board storage and memory, after the provisioning, the `BLEAgent` can be removed. ~30KB of storage and ~ 2KB of memory are saved for BSS and Data. The network credentials provided in the provisioning phase can no longer be changed via Bluetooth if this is disabled.

Reviewer: ```suggestion
To save board storage and memory, after the provisioning, the `BLEAgent` can be removed. ~30 KB of storage and ~2 KB of memory are saved for BSS and Data. The network credentials provided in the provisioning phase can no longer be changed via Bluetooth if this is disabled.
```

To disable the `BLEAgent`, just comment out or remove the lines of code that include, declare, and enable it. The final result should be the following:

```arduino
#include <ArduinoIoTCloud.h>
#include <Arduino_ConnectionHandler.h>
#include <Arduino_NetworkConfigurator.h>
//#include <configuratorAgents/agents/BLEAgent.h>
#include <configuratorAgents/agents/SerialAgent.h>
void onVariableChange();
int variable;
KVStore kvStore;
//BLEAgentClass BLEAgent;
SerialAgentClass SerialAgent;
WiFiConnectionHandler ArduinoIoTPreferredConnection; 
NetworkConfiguratorClass NetworkConfigurator(ArduinoIoTPreferredConnection);
void initProperties(){
//  NetworkConfigurator.addAgent(BLEAgent);
  NetworkConfigurator.addAgent(SerialAgent);
  NetworkConfigurator.setStorage(kvStore);
  ArduinoCloud.setConfigurator(NetworkConfigurator);
// For changing the default reset pin (2) uncomment and set your preferred pin. Use DISABLE_PIN for disabling the reset procedure.
// NetworkConfigurator.setReconfigurePin(YOUR_PIN);
  ArduinoCloud.addProperty(variable, READWRITE, ON_CHANGE, onVariableChange);
}
```

### Disable the SerialAgent

> To save board storage and memory, after the provisioning, the SerialAgent can be removed. ~1KB of storage is saved for BSS and Data. The network credentials provided in the provisioning phase can no longer be changed via USB if this is disabled.

Reviewer: ```suggestion
To save board storage and memory, after the provisioning, the SerialAgent can be removed. ~1 KB of storage is saved for BSS and Data. The network credentials provided in the provisioning phase can no longer be changed via USB if this is disabled.
```

To disable the `SerialAgent`, just comment out or remove the lines of code that include, declare, and enable it. The final result should be the following:

```arduino
#include <ArduinoIoTCloud.h>
#include <Arduino_ConnectionHandler.h>
#include <Arduino_NetworkConfigurator.h>
#include <configuratorAgents/agents/BLEAgent.h>
//#include <configuratorAgents/agents/SerialAgent.h>
void onVariableChange();
int variable;
KVStore kvStore;
BLEAgentClass BLEAgent;
//SerialAgentClass SerialAgent;
WiFiConnectionHandler ArduinoIoTPreferredConnection; 
NetworkConfiguratorClass NetworkConfigurator(ArduinoIoTPreferredConnection);
void initProperties(){
  NetworkConfigurator.addAgent(BLEAgent);
//  NetworkConfigurator.addAgent(SerialAgent);
  NetworkConfigurator.setStorage(kvStore);
  ArduinoCloud.setConfigurator(NetworkConfigurator);
// For changing the default reset pin (2) uncomment and set your preferred pin. Use DISABLE_PIN for disabling the reset procedure.
// NetworkConfigurator.setReconfigurePin(YOUR_PIN);
  ArduinoCloud.addProperty(variable, READWRITE, ON_CHANGE, onVariableChange);
}
```

### Disable the NetworkConfigurator

> To save board storage and memory, after the provisioning, the NetworkConfigurator can be removed. ~40KB of storage and ~ 2,9KB of memory are saved for BSS and Data.

Reviewer: ```suggestion
To save board storage and memory, after the provisioning, the NetworkConfigurator can be removed. ~40 kB of storage and ~2.9 kB of memory are saved for BSS and Data.
```
